rust/template/distributed_datalog/src/tcp_channel/receiver.rs

```diff
@@ -102,44 +162,74 @@ where
 
         Ok(Self {
             addr,
             fd,
             thread,
-            observer,
+            txnmux,
         })
     }
 
     /// Accept a connection (in a non-blocking manner), read data from
-    /// it, and dispatch that to the subscribed observer, if any. If no
-    /// observer is subscribed, data will be silently dropped.
+    /// it, and dispatch that to the transaction multiplexer.
     fn accept(
         listener: TcpListener,
         fd: Arc<Mutex<Fd>>,
-        mut observer: SharedObserver<OptionalObserver<ObserverBox<T, String>>>,
+        txnmux: Arc<Mutex<TxnMux<T, String>>>,
     ) -> JoinHandle<Result<(), String>> {
         spawn(move || {
-            let result = listener.accept();
-            let socket = {
-                let mut guard = fd.lock().unwrap();
-                // The user may have dropped the receiver shortly after
-                // us accepting a connection. If that is the case do not
-                // continue.
-                if let Fd::Shutdown = *guard {
-                    return Ok(());
+            let mut handles = Vec::new();
+            loop {
+                let socket = match listener.accept() {
+                    Ok((socket, _)) => socket,
+                    Err(e) => {
+                        // The user may have dropped the receiver shortly after
+                        // us accepting a connection. If that is the case do not
+                        // continue.
+                        if let Fd::Shutdown = *fd.lock().unwrap() {
+                            break;
+                        }
+                        error!("failed to accept connection: {}", e);
+                        continue;
+                    }
+                };
+
+                let passthrough = Arc::new(Mutex::new(Passthrough::new()));
+                let observable = Box::new(passthrough.clone());
+                if txnmux.lock().unwrap().add_observable(observable).is_err() {
+                    error!(
+                        "failed to register connection {} with TxnMux",
+                        socket.as_raw_fd()
+                    );
+                    continue;
                 }
 
-                let (socket, _) =
-                    result.map_err(|e| format!("failed to accept connection: {}", e))?;
-                *guard = Fd::Accepted(socket.as_raw_fd());
-                socket
-            };
+                let fd = Arc::new(Mutex::new(Fd::Active(socket.as_raw_fd())));
+                let copy = fd.clone();
+                let thread = spawn(move || Self::process(socket, copy, passthrough));
+                handles.push((thread, fd));
+            }
 
-            Self::process(socket, fd, &mut observer)
+            // We only exit above loop when the receiver is dropped and
+            // in this case we intend to stop and join all the
+            // processing threads we started.
```

Comment:
Do we need to tell the processing threads to stop? Otherwise, is there any guarantee that they will terminate sometime soon?

Reply:
We do. There is a cascading shutdown happening. When the user drops the object the "acceptor" thread's `fd` will be shutdown, which will initiate an exit of said thread. That in turn will invoke shutdown on all the individual connection sockets, unblocking the threads along. The "acceptor" thread will then wait to join these threads before exiting. `sender_cleanup` is supposed to test this case.

Reply:
> When the user drops the object the "acceptor" thread's fd will be shutdown, which will initiate an exit of said thread

Is it part of the socket API that when the accepting socket is closed, all connections it created shutdown? Can you point me to a doc?

Reply:
We discussed the matter in person. To summarize: we maintain a vector of `(socket-fd, thread handle)` pairs and when we exit the "acceptor" loop we iterate over this vector shutting down all the sockets and joining the threads. Line 220 is where the loop starts.